client/client.go

```diff
@@ -311,13 +324,110 @@ func (c *Client) Publish(topic string, options wamp.Dict, args wamp.List, kwargs
 		}
 	}
 
-	c.sess.Send(&wamp.Publish{
+	message := &wamp.Publish{
 		Request:     id,
 		Options:     options,
 		Topic:       wamp.URI(topic),
 		Arguments:   args,
 		ArgumentsKw: kwargs,
-	})
+	}
+
+	// Check and handle Payload PassThru Mode
+	// @see https://wamp-proto.org/wamp_latest_ietf.html#name-payload-passthru-mode
+	if pptScheme, _ := options[wamp.OptPPTScheme].(string); pptScheme != "" {
+		// Let's check: was ppt feature announced by broker?
+		if !c.sess.HasFeature(wamp.RoleBroker, wamp.FeaturePayloadPassthruMode) {
+			// It's protocol violation, so we need to abort connection
+			// But as we did not send anything to the router
+			// let's just err client
+			return ErrPPTNotSupported
+		}
+
+		pptSchemeIsValid := false
+		for _, v := range []*regexp.Regexp{
+			regexp.MustCompile("^wamp$"),
+			regexp.MustCompile("^mqtt$"),
+			regexp.MustCompile("^x_"),
+		} {
+			if v.MatchString(pptScheme) {
+				pptSchemeIsValid = true
+				break
+			}
+		}
+
+		if !pptSchemeIsValid {
+			return ErrPPTSchemeInvalid
+		}
+
+		// Broker supports PPT feature
+		// Let's prepare payload based on ppt_* attributes provided
+		payload := &wamp.PassthruPayload{
+			Arguments:   args,
+			ArgumentsKw: kwargs,
+		}
+
+		message.Arguments = wamp.List{payload}
+
+		// Now need to check ppt_serializer (in pair with ppt_scheme)
+		// and serialize payload with appreciate serializer
+		if pptScheme == "wamp" {
```

Comment:
wonder if that if/else could be refactored:

```
optPPTSerializer := options[wamp.OptPPTSerializer].(string)
if pptScheme == "wamp" || optPPTSerializer != "native" {
	var pptSerializer serialize.Serialization
	if pptScheme == "wamp" {
		pptSerializer = E2eeSerializers[optPPTSerializer]
	} else {
		pptSerializer = PPTSerializers[optPPTSerializer]
	}
	// TODO In future should be extended with FlatBuffers
	switch pptSerializer {
	case JSON:
		serializer = &serialize.JSONSerializer{}
	case MSGPACK:
		serializer = &serialize.MessagePackSerializer{}
	case CBOR:
		serializer = &serialize.CBORSerializer{}
	default:
		return ErrPPTSerializerInvalid
	}
	bin, err := serializer.SerializeDataItem(payload)
	if err != nil {
		return ErrSerialization
	}

	message.Arguments = wamp.List{bin}
} else {
 	message.Arguments = args
	message.ArgumentsKw = kwargs
}
```

Reply:
Also, e2ee will require access to some additional parameters of client/session... 
So I was thinking of creating independent helpers for e2ee and ppt... To adopt then e2ee as needed

Reply:
Yes, helpers are a good idea. The code is already long and nested, extracting some of it would make it more clear.